### DISCUSSION TITLE
Site-to-Site IKEv2 tunnel between strongSwan and Cisco ASA

### DISCUSSION THREAD

**Discussion by mrxanto** [strongswan/strongswan#1366] [Q&A]
Hello.

I try to establish tunnel between Cisco ASA (RESPONDER) and strongSwan (INITIATOR).
I have OpenWRT box with installed software:
<details> 
  <summary>installed software:</summary>
<pre>
strongswan - 5.8.2-2
strongswan-charon - 5.8.2-2
strongswan-ipsec - 5.8.2-2
strongswan-isakmp - 5.8.2-2
strongswan-mod-aes - 5.8.2-2
strongswan-mod-des - 5.8.2-2
strongswan-mod-gmpdh - 5.8.2-2
strongswan-mod-hmac - 5.8.2-2
strongswan-mod-kernel-netlink - 5.8.2-2
strongswan-mod-md5 - 5.8.2-2
strongswan-mod-nonce - 5.8.2-2
strongswan-mod-pubkey - 5.8.2-2
strongswan-mod-random - 5.8.2-2
strongswan-mod-sha1 - 5.8.2-2
strongswan-mod-socket-default - 5.8.2-2
strongswan-mod-stroke - 5.8.2-2
strongswan-mod-uci - 5.8.2-2
strongswan-mod-updown - 5.8.2-2
</pre>
</details>

Also they have next version of Linux kernel: 4.14.275.

On /etc/ipsec.conf I set next configuration of tunnel:
<details> 
  <summary>configuration of tunnel</summary>
<pre>
conn tunn1
 	left=%any
 	leftsubnet=10.10.20.22/32
 	right=1.2.3.4
 	rightsubnet=192.168.48.0/24
 	ikelifetime=1440m
 	lifetime=3h
 	margintime=9m
 	keyretries=%forever
 	dpdaction=restart
 	closeaction=restart
 	leftauth=psk
 	rightauth=psk
 	auto=start
 	esp=aes256-sha1
 	ike=aes256-sha2_256-prfsha1-modp2048
 	keyexchange=ikev2
 	type=tunnel
</pre>
</details>

OpenWRT box have cellular modem, which is the only internet connection for this device.
So, after power-up device IKEv2 tunnel are establish but no traffic is gone from/to device.
On Cisco ASA I see, that both (IKEv2 and IPsec) SA are present, but counters on IPsec decapsulated packets are zero:

>      #pkts encaps: 559, #pkts encrypt: 559, #pkts digest: 559
>      #pkts decaps: 0, #pkts decrypt: 0, #pkts verify: 0

<details> 
  <summary>ipsec status information: </summary>
<pre>
# ipsec status
Security Associations (1 up, 0 connecting):
        tunn1[1]: ESTABLISHED 47 minutes ago, 5.6.7.8[5.6.7.8]...1.2.3.4[1.2.3.4]
        tunn1{1}:  INSTALLED, TUNNEL, reqid 1, ESP in UDP SPIs: cf181301_i bfa99998_o
        tunn1{1}:   10.10.20.22/32 === 192.168.48.0/24
# ipsec statusall
Status of IKE charon daemon (strongSwan 5.8.2, Linux 4.14.275, mips):
  uptime: 48 minutes, since Oct 24 09:13:29 2022
  worker threads: 11 of 16 idle, 5/0/0/0 working, job queue: 0/0/0/0, scheduled: 4
  loaded plugins: charon aes des sha1 md5 random nonce pubkey gmpdh hmac kernel-netlink socket-default stroke updown
Listening IP addresses:
  5.6.7.8
  10.10.20.22
  fd14:aa41:4279::1
Connections:
        tunn1:  %any...1.2.3.4  IKEv2, dpddelay=30s
        tunn1:   local:  uses pre-shared key authentication
        tunn1:   remote: [1.2.3.4] uses pre-shared key authentication
        tunn1:   child:  10.10.20.22/32 === 192.168.48.0/24 TUNNEL, dpdaction=restart
Security Associations (1 up, 0 connecting):
        tunn1[1]: ESTABLISHED 47 minutes ago, 5.6.7.8[5.6.7.8]...1.2.3.4[1.2.3.4]
        tunn1[1]: IKEv2 SPIs: 924f22c0385e54d4_i* 088be9c0e16f8bae_r, pre-shared key reauthentication in 22 hours
        tunn1[1]: IKE proposal: AES_CBC_256/HMAC_SHA1_96/PRF_HMAC_SHA1/MODP_2048
        tunn1{1}:  INSTALLED, TUNNEL, reqid 1, ESP in UDP SPIs: cf181301_i bfa99998_o
        tunn1{1}:  AES_CBC_256/HMAC_SHA1_96, 0 bytes_i, 0 bytes_o, rekeying in 116 minutes
        tunn1{1}:   10.10.20.22/32 === 192.168.48.0/24
#
</pre>
</details>

<details> 
  <summary>ip xfrm information:</summary>
<pre>
# ip -s xfrm state
src 5.6.7.8 dst 1.2.3.4
        proto esp spi 0xbfa99998(3215563160) reqid 1(0x00000001) mode tunnel
        replay-window 0 seq 0x00000000 flag af-unspec (0x00100000)
        auth-trunc hmac(sha1) 0xbd4b867144db29b327ea8c01165fba5918d9b564 (160 bits) 96
        enc cbc(aes) 0xa15e1e95746d693d733ce4fce1eec98992dcc6811fe2968f598f82c54b0456da (256 bits)
        encap type espinudp sport 4500 dport 4500 addr 0.0.0.0
        anti-replay context: seq 0x0, oseq 0x0, bitmap 0x00000000
        lifetime config:
          limit: soft (INF)(bytes), hard (INF)(bytes)
          limit: soft (INF)(packets), hard (INF)(packets)
          expire add: soft 9850(sec), hard 10800(sec)
          expire use: soft 0(sec), hard 0(sec)
        lifetime current:
          0(bytes), 0(packets)
          add 2022-10-24 09:14:22 use -
        stats:
          replay-window 0 replay 0 failed 0
src 1.2.3.4 dst 5.6.7.8
        proto esp spi 0xcf181301(3474461441) reqid 1(0x00000001) mode tunnel
        replay-window 32 seq 0x00000000 flag af-unspec (0x00100000)
        auth-trunc hmac(sha1) 0x60c774d18013507118334f6a2b7c9efe9510b9ec (160 bits) 96
        enc cbc(aes) 0xa15bf0cf709cb6bc627698a388708467b1ef980bddbfad3491ae1d5066646110 (256 bits)
        encap type espinudp sport 4500 dport 4500 addr 0.0.0.0
        anti-replay context: seq 0x0, oseq 0x0, bitmap 0x00000000
        lifetime config:
          limit: soft (INF)(bytes), hard (INF)(bytes)
          limit: soft (INF)(packets), hard (INF)(packets)
          expire add: soft 10105(sec), hard 10800(sec)
          expire use: soft 0(sec), hard 0(sec)
        lifetime current:
          0(bytes), 0(packets)
          add 2022-10-24 09:14:21 use -
        stats:
          replay-window 0 replay 0 failed 0
# ip -s xfrm policy
src 10.10.20.22/32 dst 192.168.48.0/24 uid 0
        dir out action allow index 81 priority 371327 share any flag  (0x00000000)
        lifetime config:
          limit: soft (INF)(bytes), hard (INF)(bytes)
          limit: soft (INF)(packets), hard (INF)(packets)
          expire add: soft 0(sec), hard 0(sec)
          expire use: soft 0(sec), hard 0(sec)
        lifetime current:
          0(bytes), 0(packets)
          add 2022-10-24 09:14:22 use -
        tmpl src 5.6.7.8 dst 1.2.3.4
                proto esp spi 0xbfa99998(3215563160) reqid 1(0x00000001) mode tunnel
                level required share any
                enc-mask ffffffff auth-mask ffffffff comp-mask ffffffff
src 192.168.48.0/24 dst 10.10.20.22/32 uid 0
        dir fwd action allow index 74 priority 371327 share any flag  (0x00000000)
        lifetime config:
          limit: soft (INF)(bytes), hard (INF)(bytes)
          limit: soft (INF)(packets), hard (INF)(packets)
          expire add: soft 0(sec), hard 0(sec)
          expire use: soft 0(sec), hard 0(sec)
        lifetime current:
          0(bytes), 0(packets)
          add 2022-10-24 09:14:22 use -
        tmpl src 1.2.3.4 dst 5.6.7.8
                proto esp spi 0x00000000(0) reqid 1(0x00000001) mode tunnel
                level required share any
                enc-mask ffffffff auth-mask ffffffff comp-mask ffffffff
src 192.168.48.0/24 dst 10.10.20.22/32 uid 0
        dir in action allow index 64 priority 371327 share any flag  (0x00000000)
        lifetime config:
          limit: soft (INF)(bytes), hard (INF)(bytes)
          limit: soft (INF)(packets), hard (INF)(packets)
          expire add: soft 0(sec), hard 0(sec)
          expire use: soft 0(sec), hard 0(sec)
        lifetime current:
          0(bytes), 0(packets)
          add 2022-10-24 09:14:22 use -
        tmpl src 1.2.3.4 dst 5.6.7.8
                proto esp spi 0x00000000(0) reqid 1(0x00000001) mode tunnel
                level required share any
                enc-mask ffffffff auth-mask ffffffff comp-mask ffffffff
src 0.0.0.0/0 dst 0.0.0.0/0 uid 0
        socket in action allow index 59 priority 0 share any flag  (0x00000000)
        lifetime config:
          limit: soft 0(bytes), hard 0(bytes)
          limit: soft 0(packets), hard 0(packets)
          expire add: soft 0(sec), hard 0(sec)
          expire use: soft 0(sec), hard 0(sec)
        lifetime current:
          0(bytes), 0(packets)
          add 2022-10-21 11:44:37 use 2022-10-24 10:04:47
src 0.0.0.0/0 dst 0.0.0.0/0 uid 0
        socket out action allow index 52 priority 0 share any flag  (0x00000000)
        lifetime config:
          limit: soft 0(bytes), hard 0(bytes)
          limit: soft 0(packets), hard 0(packets)
          expire add: soft 0(sec), hard 0(sec)
          expire use: soft 0(sec), hard 0(sec)
        lifetime current:
          0(bytes), 0(packets)
          add 2022-10-21 11:44:37 use 2022-10-24 10:04:47
src 0.0.0.0/0 dst 0.0.0.0/0 uid 0
        socket in action allow index 43 priority 0 share any flag  (0x00000000)
        lifetime config:
          limit: soft 0(bytes), hard 0(bytes)
          limit: soft 0(packets), hard 0(packets)
          expire add: soft 0(sec), hard 0(sec)
          expire use: soft 0(sec), hard 0(sec)
        lifetime current:
          0(bytes), 0(packets)
          add 2022-10-21 11:44:37 use 2022-10-24 09:14:19
src 0.0.0.0/0 dst 0.0.0.0/0 uid 0
        socket out action allow index 36 priority 0 share any flag  (0x00000000)
        lifetime config:
          limit: soft 0(bytes), hard 0(bytes)
          limit: soft 0(packets), hard 0(packets)
          expire add: soft 0(sec), hard 0(sec)
          expire use: soft 0(sec), hard 0(sec)
        lifetime current:
          0(bytes), 0(packets)
          add 2022-10-21 11:44:37 use 2022-10-24 09:14:18
src ::/0 dst ::/0 uid 0
        socket in action allow index 27 priority 0 share any flag  (0x00000000)
        lifetime config:
          limit: soft 0(bytes), hard 0(bytes)
          limit: soft 0(packets), hard 0(packets)
          expire add: soft 0(sec), hard 0(sec)
          expire use: soft 0(sec), hard 0(sec)
        lifetime current:
          0(bytes), 0(packets)
          add 2022-10-21 11:44:37 use -
src ::/0 dst ::/0 uid 0
        socket out action allow index 20 priority 0 share any flag  (0x00000000)
        lifetime config:
          limit: soft 0(bytes), hard 0(bytes)
          limit: soft 0(packets), hard 0(packets)
          expire add: soft 0(sec), hard 0(sec)
          expire use: soft 0(sec), hard 0(sec)
        lifetime current:
          0(bytes), 0(packets)
          add 2022-10-21 11:44:37 use -
src ::/0 dst ::/0 uid 0
        socket in action allow index 11 priority 0 share any flag  (0x00000000)
        lifetime config:
          limit: soft 0(bytes), hard 0(bytes)
          limit: soft 0(packets), hard 0(packets)
          expire add: soft 0(sec), hard 0(sec)
          expire use: soft 0(sec), hard 0(sec)
        lifetime current:
          0(bytes), 0(packets)
          add 2022-10-21 11:44:37 use -
src ::/0 dst ::/0 uid 0
        socket out action allow index 4 priority 0 share any flag  (0x00000000)
        lifetime config:
          limit: soft 0(bytes), hard 0(bytes)
          limit: soft 0(packets), hard 0(packets)
          expire add: soft 0(sec), hard 0(sec)
          expire use: soft 0(sec), hard 0(sec)
        lifetime current:
          0(bytes), 0(packets)
          add 2022-10-21 11:44:37 use -
#
</pre>
</details>

When I run re-establish connection (via clearing SA from Cisco ASA side or just restart strongSwan daemon on OpenWRT) traffic are gone fine to/from both ends. After that I changed in /etc/ipsec.conf **auto=start** to **auto=add** and added hotplug hooks to up/down cellular modem interface with action ipsec up/down tunn1.

As I understand, when strongSwan is started, OpenWRT box is not have route to right end and there is some issue.

Maybe is there more correct way, than establish tunnel after cellular modem change their state?

[ipsec_status.txt](https://github.com/strongswan/strongswan/files/9850507/ipsec_status.txt)
[xfrm.txt](https://github.com/strongswan/strongswan/files/9850508/xfrm.txt)

**Comment by Thermi** [comment]
The problem is likely the default NAT rules. See [the docs page here](https://docs.strongswan.org/docs/5.9/howtos/forwarding.html#_general_nat_problems)

**Reply by mrxanto** [reply]
> The problem is likely the default NAT rules. See [the docs page here](https://docs.strongswan.org/docs/5.9/howtos/forwarding.html#_general_nat_problems)

I don't see differences in "iptables -t nat" between cases when ipsec pass traffic and when not.
And rule:
iptables -t nat -I POSTROUTING -m policy --pol ipsec --dir out -j ACCEPT
doesn't change anything.

Generaly speaking, I don't use NAT in this case, I just want have connection between leftsubnet (with /32 prefixes) and rightsubnet) without address translation.

**Reply by Thermi** [reply]
Can you provide the output of `ip route show table all` and `ip rule` when the VPN is up?

**Reply by mrxanto** [reply]
> Can you provide the output of `ip route show table all` and `ip rule` when the VPN is up?

<details> 
  <summary>ip route show table all:</summary>
<pre>
# ip route show table all
192.168.48.0/24 via 5.6.7.9 dev wwan0 table 220 proto static src 10.10.20.22
default via 5.6.7.9 dev wwan0 proto static src 5.6.7.8
10.10.20.0/24 dev br-lan proto kernel scope link src 10.10.20.22
5.0.0.0/8 dev wwan0 proto kernel scope link src 5.6.7.8
broadcast 10.10.20.0 dev br-lan table local proto kernel scope link src 10.10.20.22
local 10.10.20.22 dev br-lan table local proto kernel scope host src 10.10.20.22
broadcast 10.10.20.255 dev br-lan table local proto kernel scope link src 10.10.20.22
broadcast 5.0.0.0 dev wwan0 table local proto kernel scope link src 5.6.7.8
local 5.6.7.8 dev wwan0 table local proto kernel scope host src 5.6.7.8
broadcast 5.255.255.255 dev wwan0 table local proto kernel scope link src 5.6.7.8
broadcast 127.0.0.0 dev lo table local proto kernel scope link src 127.0.0.1
local 127.0.0.0/8 dev lo table local proto kernel scope host src 127.0.0.1
local 127.0.0.1 dev lo table local proto kernel scope host src 127.0.0.1
broadcast 127.255.255.255 dev lo table local proto kernel scope link src 127.0.0.1
fd14:aa41:4279::/64 dev br-lan proto static metric 1024 pref medium
unreachable fd14:aa41:4279::/48 dev lo proto static metric 2147483647 error 4294967148 pref medium
fe80::/64 dev eth0 proto kernel metric 256 pref medium
fe80::/64 dev eth0.2 proto kernel metric 256 pref medium
fe80::/64 dev br-lan proto kernel metric 256 pref medium
fe80::/64 dev wwan0 proto kernel metric 256 pref medium
local ::1 dev lo table local proto kernel metric 0 pref medium
anycast fd14:aa41:4279:: dev br-lan table local proto kernel metric 0 pref medium
local fd14:aa41:4279::1 dev br-lan table local proto kernel metric 0 pref medium
anycast fe80:: dev eth0 table local proto kernel metric 0 pref medium
anycast fe80:: dev eth0.2 table local proto kernel metric 0 pref medium
anycast fe80:: dev br-lan table local proto kernel metric 0 pref medium
anycast fe80:: dev wwan0 table local proto kernel metric 0 pref medium
local fe80::96a4:8ff:fe8e:7a20 dev eth0 table local proto kernel metric 0 pref medium
local fe80::96a4:8ff:fe8e:7a20 dev br-lan table local proto kernel metric 0 pref medium
local fe80::96a4:8ff:fe8e:7a21 dev eth0.2 table local proto kernel metric 0 pref medium
local fe80::bc61:80ff:fe24:ba91 dev wwan0 table local proto kernel metric 0 pref medium
ff00::/8 dev eth0 table local proto kernel metric 256 pref medium
ff00::/8 dev br-lan table local proto kernel metric 256 pref medium
ff00::/8 dev eth0.2 table local proto kernel metric 256 pref medium
ff00::/8 dev wwan0 table local proto kernel metric 256 pref medium
#
</pre>
</details> 

<details> 
  <summary>ip rule:</summary>
<pre>
# ip rule
0:      from all lookup local
32766:  from all lookup main
32767:  from all lookup default
#
</pre>
</details>

**Reply by Thermi** [reply]
The routing rule to look in table 220 is missing, that's the reason this is not working

**Comment by tobiasbrunner** [comment]
It might be the same issue as #861.

**Reply by mrxanto** [reply]
> It might be the same issue as #861.

At first look, indeed, option **forceencaps = yes** resolved problem. As I understand, more correctly way is update version of strongSwan, but I need some time for preparing own buildroot for compiling a newest version to my environment.
